step-automation-packages/step-automation-packages-manager/src/main/java/step/automation/packages/execution/AutomationPackageExecutor.java

```diff
@@ -0,0 +1,112 @@
+
+    private final ExecutionScheduler scheduler;
+    private final FunctionManager layeredFunctionManager;
+    private final FunctionAccessor layeredFunctionAccessor;
+    private final IsolatedAutomationPackageRepository isolatedAutomationPackageRepository;
+
+    public AutomationPackageExecutor(ExecutionScheduler scheduler,
+                                     FunctionManager layeredFunctionManager,
+                                     FunctionAccessor layeredFunctionAccessor,
+                                     IsolatedAutomationPackageRepository isolatedAutomationPackageRepository) {
+        this.scheduler = scheduler;
+        this.layeredFunctionManager = layeredFunctionManager;
+        this.layeredFunctionAccessor = layeredFunctionAccessor;
+        this.isolatedAutomationPackageRepository = isolatedAutomationPackageRepository;
+    }
+
+    public List<String> runInIsolation(InputStream automationPackage, String fileName, Map<String, String> executionParameters, ObjectEnricher objectEnricher, String userId) throws SetupFunctionException, FunctionTypeException {
+        ObjectId contextId = new ObjectId();
+
+        // prepare the isolated in-memory automation package manager with the only one automation package
+        AutomationPackageManager inMemoryPackageManager = createInMemoryAutomationPackageManager(contextId);
+        String packageId = inMemoryPackageManager.createAutomationPackage(automationPackage, fileName, objectEnricher);
+
+        List<String> executions = new ArrayList<>();
+        try {
+            isolatedAutomationPackageRepository.putContext(contextId.toString(), inMemoryPackageManager);
+
+            // TODO: filter plans
+            for (Plan plan : inMemoryPackageManager.getPackagePlans(packageId)) {
+                ExecutionParameters params = new ExecutionParameters();
+                HashMap<String, String> repositoryParameters = new HashMap<>();
+                repositoryParameters.put(IsolatedAutomationPackageRepository.REPOSITORY_PARAM_CONTEXTID, contextId.toString());
+                repositoryParameters.put(RepositoryObjectReference.PLAN_ID, plan.getId().toString());
+
+                params.setRepositoryObject(new RepositoryObjectReference(IsolatedAutomationPackageRepositoryPlugin.ISOLATED_AUTOMATION_PACKAGE, repositoryParameters));
+                params.setMode(ExecutionMode.RUN);
+                params.setDescription(plan.getRoot().getAttributes().get("name"));
+
+                if (userId != null) {
+                    params.setUserID(userId);
+                }
+                params.setCustomParameters(executionParameters);
+
+                String newExecutionId = this.scheduler.execute(params);
+                if (newExecutionId != null) {
+                    executions.add(newExecutionId);
+                }
+            }
+        } finally {
+            isolatedAutomationPackageRepository.clearContext(contextId.toString());
```

Comment:
scheduler.execute is asynchron. Shouldn't we wait for the executions to finish?

Reply:
@jeromecomte 
Yes, we should. 
But, as I understood, the REST Serivce still should return the response asynchonosly without wait for execution complete? And here I need to launch some background task to perform cleanup after execution finish (like in RemoteExecutionManager#waitForTermination). Am I right?